File: server/src/game/GameEngine.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { GameEngine } from './GameEngine.js';
import { Hero } from './Hero.js';
import { Deck } from './Deck.js';
import { Minion } from './Minion.js';
import { CardType, Rarity, HeroClass, type CardDefinition } from '@coincard/shared';
import { MAX_BOARD_SIZE } from './constants.js';

function makeMinionCard(overrides: Partial<CardDefinition> = {}): CardDefinition {
  return {
    id: `card-test-${Math.random().toString(36).slice(2)}`,
    name: 'Test Minion',
    slug: 'test-minion',
    description: 'Test.',
    type: CardType.MINION,
    rarity: Rarity.COMMON,
    manaCost: 1,
    attack: 2,
    health: 2,
    heroClass: HeroClass.NEUTRAL,
    imagePath: 'assets/images/Minions/Chilwind Yeti.png',
    effects: [],
    keywords: [],
    collectible: true,
    ...overrides,
  };
}

function makeHero(heroClass = 'MAGE'): Hero {
  return new Hero('hero_mage', 'Jaina', heroClass, 'Fireblast', 2, 'assets/images/Heros/Jaina Proudmoore.png');
}

function makeSpellCard(overrides: Partial<CardDefinition> = {}): CardDefinition {
  return {
    ...makeMinionCard(),
    id: `spell-test-${Math.random().toString(36).slice(2)}`,
    name: 'Target Spell',
    slug: 'target-spell',
    type: CardType.SPELL,
    manaCost: 0,
    attack: 0,
    health: 0,
    effects: [{ type: 'DAMAGE', value: 2, target: 'ANY_MINION' }],
    ...overrides,
  };
}

function setupEngine() {
  const engine = new GameEngine();
  const deckCards = Array.from({ length: 30 }, () => makeMinionCard());
  const game = engine.createGame('game-1', 'ABC123', [
    { playerId: 'p1', name: 'One', hero: makeHero(), deck: new Deck(deckCards) },
    { playerId: 'p2', name: 'Two', hero: makeHero(), deck: new Deck(deckCards) },
  ]);
  game.start();
  return { engine, game };
}

describe('GameEngine', () => {
  it('chia bài mở đầu: p1 3 lá, p2 4 lá', () => {
    const { game } = setupEngine();
    assert.equal(game.getPlayerById('p1').handCount, 3);
    assert.equal(game.getPlayerById('p2').handCount, 4);
  });

  it('play minion trừ mana và lên board', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p1');
    const card = player.handCards[0];
    if (!card) throw new Error('No card');
    engine.playCard('game-1', 'p1', card.id);
    assert.equal(player.boardCount, 1);
    assert.equal(player.currentMana, 0);
  });

  it('chặn play khi không đủ mana', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p1');
    player.addToHand(makeMinionCard({ manaCost: 10 }));
    const expensive = player.handCards[player.handCards.length - 1];
    if (!expensive) throw new Error('No card');
    assert.throws(() => engine.playCard('game-1', 'p1', expensive.id));
  });

  it('chặn action ngoài turn', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p2');
    const card = player.handCards[0];
    if (!card) throw new Error('No card');
    assert.throws(() => engine.playCard('game-1', 'p2', card.id));
  });

  it('minion mới summon không tấn công ngay (summoning sickness)', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p1');
    const card = player.handCards[0];
    if (!card) throw new Error('No card');
    engine.playCard('game-1', 'p1', card.id);
    const minion = player.getBoard()[0];
    if (!minion) throw new Error('No minion');
    assert.throws(() => engine.attack('game-1', 'p1', minion.instanceId, 'p2'));
  });

  it('end turn đổi active player và tăng mana', () => {
    const { engine, game } = setupEngine();
    engine.endTurn('game-1', 'p1');
    assert.equal(game.getActivePlayerId(), 'p2');
    assert.equal(game.getPlayerById('p2').currentMaxMana, 1);
  });

  it('minion tấn hero gây damage và kết thúc game khi HP <= 0', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    p2.heroState.takeDamage(29); // còn 1 HP
    const card = p1.handCards[0];
    if (!card) throw new Error('No card');
    engine.playCard('game-1', 'p1', card.id);
    const minion = p1.getBoard()[0];
    if (!minion) throw new Error('No minion');
    // Hết summoning sickness: end turn 2 lần để minion sẵn sàng
    engine.endTurn('game-1', 'p1');
    engine.endTurn('game-1', 'p2');
    engine.attack('game-1', 'p1', minion.instanceId, 'p2');
    assert.equal(game.isFinished(), true);
    assert.equal(game.getWinnerId(), 'p1');
  });

  it('minion vs minion cùng gây damage, con chết bị xóa', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    // p1 summon
    const c1 = p1.handCards[0];
    if (!c1) throw new Error('No card');
    engine.playCard('game-1', 'p1', c1.id);
    engine.endTurn('game-1', 'p1');
    // Thêm 1/1 vào tay p2 sau khi draw đầu turn để chắc chắn lấy đúng lá
    p2.addToHand(makeMinionCard({ attack: 1, health: 1 }));
    // p2 summon con 1/1
    const c2 = p2.handCards[p2.handCards.length - 1];
    if (!c2) throw new Error('No card');
    assert.equal(c2.attack, 1);
    assert.equal(c2.health, 1);
    engine.playCard('game-1', 'p2', c2.id);
    engine.endTurn('game-1', 'p2');
    const attacker = p1.getBoard()[0];
    const defender = p2.getBoard()[0];
    if (!attacker || !defender) throw new Error('No minions');
    engine.attack('game-1', 'p1', attacker.instanceId, defender.instanceId);
    assert.equal(p2.boardCount, 0); // 1/1 chết
    assert.equal(p1.boardCount, 1); // 2/2 còn 1 HP
  });

  it(`board tối đa ${MAX_BOARD_SIZE} minion`, () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    // Tay giới hạn 6 lá: xả tay mở đầu, đánh theo đợt để đủ board 7.
    for (const c of [...p1.handCards]) {
      p1.removeFromHand(c.id);
    }
    p1.refillMana();
    let played = 0;
    while (played < MAX_BOARD_SIZE) {
      const need = Math.min(6 - p1.handCount, MAX_BOARD_SIZE - played);
      for (let i = 0; i < need; i++) {
        p1.addToHand(makeMinionCard({ manaCost: 0 }));
      }
      const batch = p1.handCount;
      for (let i = 0; i < batch && played < MAX_BOARD_SIZE; i++) {
        const card = p1.handCards[p1.handCards.length - 1];
        if (!card) throw new Error('No card');
        engine.playCard('game-1', 'p1', card.id);
        played++;
      }
    }
    assert.equal(p1.boardCount, MAX_BOARD_SIZE);
    p1.addToHand(makeMinionCard({ manaCost: 0 }));
    const extra = p1.handCards[p1.handCards.length - 1];
    if (!extra) throw new Error('No card');
    assert.throws(() => engine.playCard('game-1', 'p1', extra.id));
  });

  it('board đầy: action bị reject không tăng mana/maxMana và không mất bài', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p1');
    for (const c of [...player.handCards]) {
      player.removeFromHand(c.id);
    }
    for (let i = 0; i < MAX_BOARD_SIZE; i++) {
      const card = makeMinionCard({ manaCost: 0 });
      player.addToHand(card);
      engine.playCard('game-1', 'p1', card.id);
    }
    const rejected = makeMinionCard({ manaCost: 0 });
    player.addToHand(rejected);
    const manaBefore = player.currentMana;
    const maxManaBefore = player.currentMaxMana;
    const handBefore = player.handCount;
    assert.throws(() => engine.playCard('game-1', 'p1', rejected.id));
    assert.equal(player.currentMana, manaBefore);
    assert.equal(player.currentMaxMana, maxManaBefore);
    assert.equal(player.handCount, handBefore);
    assert.equal(player.findCardInHand(rejected.id)?.id, rejected.id);
  });

  it('spell không có target: reject nguyên tử, không mất mana/bài', () => {
    const { engine, game } = setupEngine();
    const player = game.getPlayerById('p1');
    const spell = makeSpellCard();
    player.addToHand(spell);
    const manaBefore = player.currentMana;
    const handBefore = player.handCount;
    assert.throws(() => engine.playCard('game-1', 'p1', spell.id));
    assert.equal(player.currentMana, manaBefore);
    assert.equal(player.handCount, handBefore);
    assert.equal(player.findCardInHand(spell.id)?.id, spell.id);
  });

  it('Paladin hero power board đầy: không mất mana', () => {
    const engine = new GameEngine();
    const cards = Array.from({ length: 30 }, () => makeMinionCard({ manaCost: 0 }));
    const game = engine.createGame('paladin-game', 'PAL123', [
      { playerId: 'p1', name: 'One', hero: makeHero('PALADIN'), deck: new Deck(cards) },
      { playerId: 'p2', name: 'Two', hero: makeHero(), deck: new Deck(cards) },
    ]);
    game.start();
    const player = game.getPlayerById('p1');
    for (const c of [...player.handCards]) {
      player.removeFromHand(c.id);
    }
    for (let i = 0; i < MAX_BOARD_SIZE; i++) {
      const card = makeMinionCard({ manaCost: 0 });
      player.addToHand(card);
      engine.playCard('paladin-game', 'p1', card.id);
    }
    const manaBefore = player.currentMana;
    assert.throws(() => engine.useHeroPower('paladin-game', 'p1'));
    assert.equal(player.currentMana, manaBefore);
    assert.equal(player.boardCount, MAX_BOARD_SIZE);
  });

  it('finish idempotent: winner không bị ghi đè', () => {
    const { game } = setupEngine();
    game.finish('p1');
    game.finish('p2');
    assert.equal(game.getWinnerId(), 'p1');
  });

  it('tay giới hạn 6 lá: addToHand quá 6 bị burn', () => {
    const { game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    for (const c of [...p1.handCards]) {
      p1.removeFromHand(c.id);
    }
    for (let i = 0; i < 8; i++) {
      p1.addToHand(makeMinionCard({ manaCost: 0 }));
    }
    assert.equal(p1.handCount, 6);
  });

  it('rút bài thủ công: mỗi turn 1 lá, turn sau được rút tiếp', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const before = p1.handCount;
    engine.drawCard('game-1', 'p1');
    assert.equal(p1.handCount, before + 1);
    assert.throws(() => engine.drawCard('game-1', 'p1'));
    engine.endTurn('game-1', 'p1');
    engine.endTurn('game-1', 'p2');
    const before2 = p1.handCount;
    engine.drawCard('game-1', 'p1');
    assert.equal(p1.handCount, before2 + 1);
  });

  it('rút bài khi tay đầy hoặc deck hết bị reject', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    for (const c of [...p1.handCards]) {
      p1.removeFromHand(c.id);
    }
    for (let i = 0; i < 6; i++) {
      p1.addToHand(makeMinionCard({ manaCost: 0 }));
    }
    assert.throws(() => engine.drawCard('game-1', 'p1'));
  });

  it('spell ENEMY_MINION: reject quái mình (nguyên tử), cho phép quái địch', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    p1.summonMinion(new Minion('own-1', 'c', 'Own', 2, 5, 'p1', false, 'img'));
    p2.summonMinion(new Minion('foe-1', 'c', 'Foe', 2, 5, 'p2', false, 'img'));
    const curse = makeSpellCard({
      manaCost: 0,
      effects: [{ type: 'DAMAGE', value: 2, target: 'ENEMY_MINION' }],
    });
    p1.addToHand(curse);
    const manaBefore = p1.currentMana;
    const handBefore = p1.handCount;
    assert.throws(() => engine.playCard('game-1', 'p1', curse.id, 'own-1'));
    assert.equal(p1.handCount, handBefore);
    assert.equal(p1.currentMana, manaBefore);
    assert.equal(p1.findMinion('own-1')?.currentHealth, 5);
    engine.playCard('game-1', 'p1', curse.id, 'foe-1');
    assert.equal(p2.findMinion('foe-1')?.currentHealth, 3);
  });

  it('Siphon Soul: cast được — destroy đúng target, heal hero mình', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    p2.summonMinion(new Minion('big-1', 'c', 'Big', 6, 6, 'p2', false, 'img'));
    p1.heroState.takeDamage(10);
    const siphon = makeSpellCard({
      manaCost: 0,
      effects: [
        { type: 'DESTROY', value: 0, target: 'ANY_MINION' },
        { type: 'HEAL', value: 3, target: 'FRIENDLY_HERO' },
      ],
    });
    p1.addToHand(siphon);
    engine.playCard('game-1', 'p1', siphon.id, 'big-1');
    assert.equal(p2.boardCount, 0);
    assert.equal(p1.heroState.currentHealth, 23);
  });

  it('attack: chặn dùng quái địch + tự đánh mình, fail không mất lượt', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    const attacker = new Minion('atk-1', 'c', 'Atk', 3, 3, 'p1', true, 'img');
    const mate = new Minion('mate-1', 'c', 'Mate', 2, 2, 'p1', false, 'img');
    const foe = new Minion('foe-2', 'c', 'Foe', 2, 2, 'p2', true, 'img');
    p1.summonMinion(attacker);
    p1.summonMinion(mate);
    p2.summonMinion(foe);
    assert.throws(() => engine.attack('game-1', 'p1', 'foe-2', 'p2'));
    assert.throws(() => engine.attack('game-1', 'p1', 'atk-1', 'p1'));
    assert.throws(() => engine.attack('game-1', 'p1', 'atk-1', 'mate-1'));
    assert.equal(attacker.canAttack, true);
    engine.attack('game-1', 'p1', 'atk-1', 'p2');
    assert.equal(p2.heroState.currentHealth, 27);
    assert.equal(attacker.canAttack, false);
  });

  it('Taunt: chặn đánh hero/quái thường, cho đánh Taunt', () => {
    const { engine, game } = setupEngine();
    const p1 = game.getPlayerById('p1');
    const p2 = game.getPlayerById('p2');
    const attacker = new Minion('atk-2', 'c', 'Atk', 3, 5, 'p1', true, 'img');
    p1.summonMinion(attacker);
    p2.summonMinion(new Minion('taunt-1', 'c', 'Taunt', 2, 4, 'p2', false, 'img', true));
    p2.summonMinion(new Minion('plain-1', 'c', 'Plain', 2, 4, 'p2', false, 'img'));
    assert.throws(() => engine.attack('game-1', 'p1', 'atk-2', 'p2'));
    assert.throws(() => engine.attack('game-1', 'p1', 'atk-2', 'plain-1'));
    assert.equal(attacker.canAttack, true);
    engine.attack('game-1', 'p1', 'atk-2', 'taunt-1');
    assert.equal(p2.findMinion('taunt-1')?.currentHealth, 1);
  });

  it('tay đầu luôn có lá rẻ (opening guarantee)', () => {
    for (let i = 0; i < 20; i++) {
      const { game } = setupEngine();
      const cheap1 = game.getPlayerById('p1').handCards.some((c) => c.manaCost <= 2);
      const cheap2 = game.getPlayerById('p2').handCards.some((c) => c.manaCost <= 2);
      assert.equal(cheap1, true);
      assert.equal(cheap2, true);
      assert.equal(game.getPlayerById('p1').handCount, 3);
      assert.equal(game.getPlayerById('p2').handCount, 4);
    }
  });
});

File: server-java/src/test/java/vn/coincard/server/game/GameEngineTest.java
package vn.coincard.server.game;

import static org.junit.jupiter.api.Assertions.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/** Port of the core GameEngine.test.ts cases. */
class GameEngineTest {
  private GameEngine engine;

  static CardTypes.CardDefinition minionCard(String id, int cost, int atk, int hp, String... keywords) {
    return new CardTypes.CardDefinition(id, "Test Minion", "test-minion", "Test.", "MINION",
        "COMMON", cost, atk, hp, "NEUTRAL", "assets/images/Minions/Chilwind Yeti.png",
        List.of(), List.of(keywords), true);
  }

  static CardTypes.CardDefinition spellCard(String id, int cost, CardTypes.EffectDefinition... effects) {
    return new CardTypes.CardDefinition(id, "Target Spell", "target-spell", "Test.", "SPELL",
        "COMMON", cost, 0, 0, "NEUTRAL", "assets/images/Minions/Chilwind Yeti.png",
        List.of(effects), List.of(), true);
  }

  static Hero mage() {
    return new Hero("hero_mage", "Jaina", "MAGE", "Fireblast", 2,
        "assets/images/Heros/Jaina Proudmoore.png");
  }

  static Deck deckOf(int n) {
    List<CardTypes.CardDefinition> cards = new ArrayList<>();
    for (int i = 0; i < n; i++) cards.add(minionCard("c-" + i + "-" + UUID.randomUUID(), 1, 2, 2));
    return new Deck(cards);
  }

  Game setup() {
    engine = new GameEngine();
    Game game = engine.createGame("game-1", "ABC123",
        new GameEngine.PlayerInit("p1", "One", mage(), deckOf(30)),
        new GameEngine.PlayerInit("p2", "Two", mage(), deckOf(30)));
    game.start();
    return game;
  }

  @Test
  void openingHands() {
    Game game = setup();
    assertEquals(3, game.getPlayerById("p1").handCount());
    assertEquals(4, game.getPlayerById("p2").handCount());
  }

  @Test
  void playMinionSpendsManaAndBoards() {
    Game game = setup();
    Player p = game.getPlayerById("p1");
    CardTypes.CardDefinition card = p.handCards().get(0);
    engine.playCard("game-1", "p1", card.id(), null);
    assertEquals(1, p.boardCount());
    assertEquals(0, p.currentMana());
  }

  @Test
  void rejectsWithoutMana() {
    Game game = setup();
    Player p = game.getPlayerById("p1");
    CardTypes.CardDefinition expensive = minionCard("big", 10, 8, 8);
    p.addToHand(expensive);
    assertThrows(GameException.NotEnoughMana.class,
        () -> engine.playCard("game-1", "p1", "big", null));
  }

  @Test
  void rejectsOffTurn() {
    Game game = setup();
    Player p2 = game.getPlayerById("p2");
    CardTypes.CardDefinition card = p2.handCards().get(0);
    assertThrows(GameException.NotPlayerTurn.class,
        () -> engine.playCard("game-1", "p2", card.id(), null));
  }

  @Test
  void summoningSickness() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    CardTypes.CardDefinition card = p1.handCards().get(0);
    engine.playCard("game-1", "p1", card.id(), null);
    Minion minion = p1.getBoard().get(0);
    assertThrows(GameException.class,
        () -> engine.attack("game-1", "p1", minion.instanceId, "p2"));
  }

  @Test
  void endTurnSwitchesAndRampsMana() {
    Game game = setup();
    engine.endTurn("game-1", "p1");
    assertEquals("p2", game.getActivePlayerId());
    assertEquals(1, game.getPlayerById("p2").currentMaxMana());
  }

  @Test
  void minionKillsHeroAndFinishes() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    p2.heroState().takeDamage(29);
    CardTypes.CardDefinition card = p1.handCards().get(0);
    engine.playCard("game-1", "p1", card.id(), null);
    Minion minion = p1.getBoard().get(0);
    engine.endTurn("game-1", "p1");
    engine.endTurn("game-1", "p2");
    engine.attack("game-1", "p1", minion.instanceId, "p2");
    assertTrue(game.isFinished());
    assertEquals("p1", game.getWinnerId());
  }

  @Test
  void boardFullRejectIsAtomic() {
    Game game = setup();
    Player p = game.getPlayerById("p1");
    for (CardTypes.CardDefinition c : new ArrayList<>(p.handCards())) {
      p.removeFromHand(c.id());
    }
    p.refillMana();
    int played = 0;
    while (played < Constants.MAX_BOARD_SIZE) {
      int need = Math.min(6 - p.handCount(), Constants.MAX_BOARD_SIZE - played);
      for (int i = 0; i < need; i++) p.addToHand(minionCard("z-" + played + "-" + i, 0, 1, 1));
      int batch = p.handCount();
      for (int i = 0; i < batch && played < Constants.MAX_BOARD_SIZE; i++) {
        CardTypes.CardDefinition c = p.handCards().get(p.handCount() - 1);
        engine.playCard("game-1", "p1", c.id(), null);
        played++;
      }
    }
    assertEquals(Constants.MAX_BOARD_SIZE, p.boardCount());
    CardTypes.CardDefinition extra = minionCard("extra", 0, 1, 1);
    p.addToHand(extra);
    int manaBefore = p.currentMana(), maxBefore = p.currentMaxMana(), handBefore = p.handCount();
    assertThrows(GameException.BoardFull.class,
        () -> engine.playCard("game-1", "p1", "extra", null));
    assertEquals(manaBefore, p.currentMana());
    assertEquals(maxBefore, p.currentMaxMana());
    assertEquals(handBefore, p.handCount());
    assertNotNull(p.findCardInHand("extra"));
  }

  @Test
  void spellWithoutTargetRejectsAtomically() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    CardTypes.CardDefinition spell = spellCard("s1", 0,
        new CardTypes.EffectDefinition("DAMAGE", 2, "ANY_MINION", null, null, null));
    p1.addToHand(spell);
    int manaBefore = p1.currentMana(), handBefore = p1.handCount();
    assertThrows(GameException.class, () -> engine.playCard("game-1", "p1", "s1", null));
    assertEquals(manaBefore, p1.currentMana());
    assertEquals(handBefore, p1.handCount());
    assertNotNull(p1.findCardInHand("s1"));
  }

  @Test
  void enemyOnlySpellRejectsOwnMinion() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    p1.summonMinion(new Minion("own-1", "c", "Own", 2, 5, "p1", false, "img"));
    p2.summonMinion(new Minion("foe-1", "c", "Foe", 2, 5, "p2", false, "img"));
    CardTypes.CardDefinition curse = spellCard("curse", 0,
        new CardTypes.EffectDefinition("DAMAGE", 2, "ENEMY_MINION", null, null, null));
    p1.addToHand(curse);
    int handBefore = p1.handCount();
    assertThrows(GameException.class, () -> engine.playCard("game-1", "p1", "curse", "own-1"));
    assertEquals(handBefore, p1.handCount());
    assertEquals(5, p1.findMinion("own-1").currentHealth());
    engine.playCard("game-1", "p1", "curse", "foe-1");
    assertEquals(3, p2.findMinion("foe-1").currentHealth());
  }

  @Test
  void siphonSoulCastsAndHeals() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    p2.summonMinion(new Minion("big-1", "c", "Big", 6, 6, "p2", false, "img"));
    p1.heroState().takeDamage(10);
    CardTypes.CardDefinition siphon = spellCard("siphon", 0,
        new CardTypes.EffectDefinition("DESTROY", 0, "ANY_MINION", null, null, null),
        new CardTypes.EffectDefinition("HEAL", 3, "FRIENDLY_HERO", null, null, null));
    p1.addToHand(siphon);
    engine.playCard("game-1", "p1", "siphon", "big-1");
    assertEquals(0, p2.boardCount());
    assertEquals(23, p1.heroState().currentHealth());
  }

  @Test
  void attackGuardsAndTaunt() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    Minion attacker = new Minion("atk-1", "c", "Atk", 3, 3, "p1", true, "img");
    p1.summonMinion(attacker);
    p2.summonMinion(new Minion("foe-2", "c", "Foe", 2, 2, "p2", true, "img"));
    assertThrows(GameException.class, () -> engine.attack("game-1", "p1", "foe-2", "p2"));
    assertThrows(GameException.class, () -> engine.attack("game-1", "p1", "atk-1", "p1"));
    p2.removeMinion("foe-2");
    engine.attack("game-1", "p1", "atk-1", "p2");
    assertEquals(27, p2.heroState().currentHealth());

    Minion attacker2 = new Minion("atk-2", "c", "Atk", 3, 5, "p1", true, "img");
    p1.summonMinion(attacker2);
    p2.summonMinion(new Minion("taunt-1", "c", "Taunt", 2, 4, "p2", false, "img", true));
    p2.summonMinion(new Minion("plain-1", "c", "Plain", 2, 4, "p2", false, "img"));
    assertThrows(GameException.MinionsBlockHero.class,
        () -> engine.attack("game-1", "p1", "atk-2", "p2"));
    assertThrows(GameException.TauntRequired.class,
        () -> engine.attack("game-1", "p1", "atk-2", "plain-1"));
    engine.attack("game-1", "p1", "atk-2", "taunt-1");
    assertEquals(1, p2.findMinion("taunt-1").currentHealth());
  }

  @Test
  void mustClearMinionsBeforeHero() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    Minion a = new Minion("a3", "c", "A", 3, 3, "p1", true, "img");
    Minion b = new Minion("b3", "c", "B", 3, 3, "p1", true, "img");
    p1.summonMinion(a);
    p1.summonMinion(b);
    p2.summonMinion(new Minion("f4", "c", "F", 2, 2, "p2", false, "img"));
    GameException ex = assertThrows(GameException.class,
        () -> engine.attack("game-1", "p1", "a3", "p2"));
    assertEquals("MINIONS_BLOCK_HERO", ex.code());
    assertTrue(a.canAttack());
    engine.attack("game-1", "p1", "a3", "f4");
    assertEquals(0, p2.boardCount());
    engine.attack("game-1", "p1", "b3", "p2");
    assertEquals(27, p2.heroState().currentHealth());
  }

  @Test
  void handLimitAndManualDraw() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    for (CardTypes.CardDefinition c : new ArrayList<>(p1.handCards())) p1.removeFromHand(c.id());
    for (int i = 0; i < 8; i++) p1.addToHand(minionCard("h" + i, 0, 1, 1));
    assertEquals(6, p1.handCount());
    assertThrows(GameException.HandFull.class, () -> engine.drawCard("game-1", "p1"));
  }

  @Test
  void bothHeroesDieIsDraw() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    p1.heroState().takeDamage(29);
    p2.heroState().takeDamage(29);
    CardTypes.CardDefinition boom = spellCard("boom", 0,
        new CardTypes.EffectDefinition("DAMAGE", 30, "ENEMY_HERO", null, null, null),
        new CardTypes.EffectDefinition("DAMAGE", 30, "FRIENDLY_HERO", null, null, null));
    p1.addToHand(boom);
    engine.playCard("game-1", "p1", "boom", null);
    assertTrue(game.isFinished());
    assertNull(game.getWinnerId());
  }

  @Test
  void statsCountActualDamage() {
    Game game = setup();
    Player p1 = game.getPlayerById("p1");
    Player p2 = game.getPlayerById("p2");
    CardTypes.CardDefinition charger = minionCard("chg", 0, 3, 3, "CHARGE");
    p1.addToHand(charger);
    engine.playCard("game-1", "p1", "chg", null);
    assertEquals(1, p1.cardsPlayed());
    Minion m = p1.getBoard().stream().filter(x -> x.cardId.equals("chg")).findFirst().orElseThrow();
    engine.attack("game-1", "p1", m.instanceId, "p2");
    assertEquals(3, p1.damageDealt());
    assertEquals(27, p2.heroState().currentHealth());
  }
}
